### PR TITLE
Remove merchant login/logout calls

### LINKED ISSUE #29191: Remove merchant login/logout calls around createSimpleProduct()
Originally `createSimpleProduct()` used the dashboard interface to create the product. Now it uses the `api` package so being logged in as the merchant isn't required.

Most of the tests that use `createSimpleProduct()` still follow the pattern of the original method and log in the merchant. We should remove all of these to improve the speed of the tests.

Noting as well that the shopper tests that log in do not log out and are relying on the merchant login to log the shopper out from the previous test. We should ensure that all tests end with no user logged in.

### PR DESCRIPTION
**Pull request by rodelgc**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #29191 .

This PR removes redundant `merchant.login()` and `merchant.logout()` around the `createSimpleProduct()` function calls throughout the test suite.

Also, this replaces `switchUserToTest()` with `merchant.logout()` in the `jest.setup.js` file. This is because  `switchUserToTest()` no longer seems to log the merchant out.

### How to test the changes in this Pull Request:

1. `cd` into the `woocommerce` repo and bring the e2e environment up.
2. Run the setup-onboarding test:

    ```
    npx wc-e2e test:e2e tests/e2e/specs/activate-and-setup/setup-onboarding.js
    ```
2. Run all the tests affected by this PR. You can use these commands to run them individually:

    ```
    # This test always fails in my headless runs
    # But always passes in non-headless mode
    npx wc-e2e test:e2e-dev tests/e2e/specs/wp-admin/order-coupon.test.js

    npx wc-e2e test:e2e tests/e2e/specs/wp-admin/order-refund.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/wp-admin/product-edit.test.js
    npx wc-e2e test:e2e tests/e2e/specs/wp-admin/product-search.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/wp-admin/create-shipping-zones.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/front-end/cart-coupons.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/front-end/cart-redirection.cart.js
    npx wc-e2e test:e2e tests/e2e/specs/front-end/cart-begin.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/front-end/checkout-coupons.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/front-end/checkout-begin.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/front-end/my-account-pay-order.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/front-end/order-email-receiving.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/front-end/single-product-page.test.js

    # These tests rely on payment settings being setup first
    npx wc-e2e test:e2e tests/e2e/specs/front-end/checkout-login-account.test.js
    npx wc-e2e test:e2e tests/e2e/specs/wp-admin/order-customer-payment-page.test.js 
    npx wc-e2e test:e2e tests/e2e/specs/front-end/checkout-create-account.test.js
    ```
1. All these tests should pass

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Remove unnecessary `merchant.login()` and `merchant.logout()` calls around `createSimpleProduct()`
> Replace  `switchUserToTest()` with `merchant.logout()` in the `jest.setup.js` file

### CODE REVIEW

File: tests/e2e/core-tests/specs/shopper/front-end-order-email-receiving.test.js [hunk 1]
```diff
@@ -29,7 +29,6 @@ const {
 const runOrderEmailReceivingTest = () => {
 	describe('Shopper Order Email Receiving', () => {
 		beforeAll(async () => {
-			await merchant.login();
```
Comment (zhongruige) [inline tests/e2e/core-tests/specs/shopper/front-end-order-email-receiving.test.js:32]: We may want to consider keeping this one here. The `await deleteAllEmailLogs();` method below does not log the merchant in:

```
const deleteAllEmailLogs = async () => {
	await merchant.openEmailLog();

	// Make sure we have emails to delete. If we don't, this selector will return null.
	if ( await page.$( '#bulk-action-selector-top' ) !== null ) {
		await setCheckbox( '#cb-select-all-1' );
		await expect( page ).toSelect( '#bulk-action-selector-top', 'Delete' );
		await Promise.all( [
			page.click( '#doaction' ),
			page.waitForNavigation( { waitUntil: 'networkidle0' } ),
		] );
	}
};
```

File: tests/e2e/core-tests/specs/shopper/front-end-single-product.test.js [hunk 2]
```diff
@@ -20,7 +20,6 @@ const simpleProductName = config.get( 'products.simple.name' );
 const runSingleProductPageTest = () => {
 	describe('Single Product Page', () => {
 		beforeAll(async () => {
-			await merchant.login();
 			simplePostIdValue = await createSimpleProduct();
 			await merchant.logout();
```
Comment (zhongruige) [inline tests/e2e/core-tests/specs/shopper/front-end-single-product.test.js:24]: We should be able to remove this line as well.

File: tests/e2e/core-tests/specs/shopper/front-end-my-account-pay-order.test.js [hunk 3]
```diff
@@ -17,7 +17,6 @@ const simpleProductName = config.get( 'products.simple.name' );
 const runMyAccountPayOrderTest = () => {
 	describe('Customer can pay for their order through My Account', () => {
 		beforeAll(async () => {
-			await merchant.login();
 			simplePostIdValue = await createSimpleProduct();
 			await merchant.logout();
```
Comment (zhongruige) [inline tests/e2e/core-tests/specs/shopper/front-end-my-account-pay-order.test.js:21]: We should be safe to remove this logout() call here.

**Comment by rrennick** [comment]
@rodel-calasagsag The local jest `beforeAll()` setup function logs in as the merchant. It was written to log out the merchant but it looks like that may not be working. 

We probably need to change https://github.com/woocommerce/woocommerce/blob/trunk/tests/e2e/config/jest.setup.js#L38 to `await merchant.logout()`. Doing that may break other tests that have been working because the merchant was still logged in at the beginning of the test. For those, add the `await merchant.login()` to the test `beforeAll()` as needed.

File: tests/e2e/core-tests/specs/shopper/front-end-order-email-receiving.test.js (hunk 1 above)
Comment (rodel-calasagsag) [reply tests/e2e/core-tests/specs/shopper/front-end-order-email-receiving.test.js]: Thanks! I re-added the `merchant.login()` call.

FYI: The reason why the test was passing even without the merchant login is that the jest setup config doesn't seem to log the merchant out, as mentioned by @rrennick below

**Comment by rodel-calasagsag** [comment]
> The local jest beforeAll() setup function logs in as the merchant. It was written to log out the merchant but it looks like that may not be working.

@rrennick Thanks! Makes sense now. I'll replace that line with merchant logout now, and fix the affected tests

File: tests/e2e/core-tests/specs/shopper/front-end-single-product.test.js (hunk 2 above)
Comment (rodel-calasagsag) [reply tests/e2e/core-tests/specs/shopper/front-end-single-product.test.js]: Done removing this unnecessary merchant logout.

File: tests/e2e/core-tests/specs/shopper/front-end-my-account-pay-order.test.js (hunk 3 above)
Comment (rodel-calasagsag) [reply tests/e2e/core-tests/specs/shopper/front-end-my-account-pay-order.test.js]: Done removing this unnecessary merchant logout.

File: tests/e2e/core-tests/specs/shopper/front-end-single-product.test.js (hunk 2 above)
Comment (zhongruige) [reply tests/e2e/core-tests/specs/shopper/front-end-single-product.test.js]: Thanks!

File: tests/e2e/core-tests/specs/shopper/front-end-my-account-pay-order.test.js (hunk 3 above)
Comment (zhongruige) [reply tests/e2e/core-tests/specs/shopper/front-end-my-account-pay-order.test.js]: Thanks for taking care of this!

**Comment by rodel-calasagsag** [comment]
This PR is ready for another round of review 🙂 

### Notable changes since the last review:

I noticed that merchant/shopper logins were being preserved across different tests (which is not the intended behavior) because of 2 reasons:

1. `tests/e2e/config/jest.setup.js` was not logging the merchant out when there are no existing posts to trash. Adding a `merchant.logout()` inside the `if()` condition there will solve this, but I went in a different route: I used [this WordPress API endpoint](https://developer.wordpress.org/rest-api/reference/posts/#delete-a-post) to trash existing posts. I thought that doing it via the API would be more efficient than via the UI. Let me know if this is fine. Here's the [commit](https://github.com/woocommerce/woocommerce/pull/29939/commits/8a7f1856529a9e0dec08eec16c54edf1ba86db5d) where I did this change.

2. I noticed that when I run the e2e tests in non-headless mode, I see two Chromium tabs. Only one of these tabs is closed when a test finishes. So when the next test launches up a new browser tab, I think what happens is that the browser cache and cookies are being preserved even though Chromium is in incognito mode, thereby preserving the logged-in session of the merchant/shopper from the previous test. The solution I did here is to clear the browser's cookies and cache entirely using Chrome DevTools Protocol after every test, and then added the necessary merchant and shopper logins to the affected tests. Here's the [commit](https://github.com/woocommerce/woocommerce/pull/29939/commits/072af834b9e1480e94368a9e2870f7333e3b0312).

**Comment by rrennick** [comment]
> but I went in a different route: I used this WordPress API endpoint to trash existing posts. 

@rodel-calasagsag The reason I left the browser trashing the posts is that it initializes the browser for the browser setup calls that follow it. You solved this by adding the goto login page. Nice fix for this.

File: tests/e2e/config/jest.setup.js [hunk 4]
```diff
@@ -1,50 +1,52 @@
...
+	await page.goto(WP_ADMIN_LOGIN);
 	await clearLocalStorage();
-	await setBrowserViewport( 'large' );
-} );
+	await setBrowserViewport('large');
+});
+
+// Clear browser cookies and cache using DevTools.
+// This is to ensure that each test ends with no user logged in.
+afterAll(async () => {
```
Comment (zhongruige) [inline tests/e2e/config/jest.setup.js:48]: Nice solution here!

**Review by zhongruige — APPROVED** [review]
Great work on this @rodel-calasagsag! I appreciated the updates for clearing the cookies and cache, as well as using the API instead of the UI to delete existing posts. These ran great for me locally, too:

```
Test Suites: 39 passed, 39 total
Tests:       3 skipped, 163 passed, 166 total
Snapshots:   0 total
Time:        746.965s
Ran all test suites.
Jest exit code: 0
```

**Pull request merged**